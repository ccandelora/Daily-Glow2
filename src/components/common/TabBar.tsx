import React from 'react';
import { View, TouchableOpacity, StyleSheet } from 'react-native';
import { Typography } from './Typography';
import theme from '@/constants/theme';
import type { BottomTabBarProps } from '@react-navigation/bottom-tabs';

interface TabItem {
  key: string;
  label: string;
  icon: string;
}

interface TabBarProps extends BottomTabBarProps {
  tabs: TabItem[];
}

export const TabBar: React.FC<TabBarProps> = ({
  state,
  navigation,
  tabs,
}) => {
  const activeTab = state.routeNames[state.index];

  const handlePress = (key: string) => {
    const event = navigation.emit({
      type: 'tabPress',
      target: key,
      canPreventDefault: true,
    });

    if (!event.defaultPrevented) {
      navigation.navigate(key);
    }
  };

  return (
    <View style={styles.container}>
      {tabs.map((tab) => (
        <TouchableOpacity
          key={tab.key}
          style={[
            styles.tab,
            activeTab === tab.key && styles.activeTab,
          ]}
          onPress={() => handlePress(tab.key)}
        >
          <Typography
            style={styles.icon}
            color={activeTab === tab.key ? theme.COLORS.primary.green : theme.COLORS.ui.textSecondary}
          >
            {tab.icon}
          </Typography>
          <Typography
            variant="caption"
            color={activeTab === tab.key ? theme.COLORS.primary.green : theme.COLORS.ui.textSecondary}
            numberOfLines={1}
          >
            {tab.label}
          </Typography>
        </TouchableOpacity>
      ))}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    backgroundColor: theme.COLORS.ui.card,
    paddingBottom: theme.SPACING.xl,
    paddingTop: theme.SPACING.md,
    borderTopWidth: 1,
    borderTopColor: theme.COLORS.ui.border,
  },
  tab: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    paddingHorizontal: theme.SPACING.xs,
  },
  activeTab: {
    backgroundColor: 'transparent',
  },
  icon: {
    fontSize: 22,
    marginBottom: 4,
  },
}); 