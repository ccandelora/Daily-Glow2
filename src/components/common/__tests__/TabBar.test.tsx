import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { TabBar } from '../TabBar';

// Mock navigation state and functions
const createMockNavigation = () => ({
  emit: jest.fn(() => ({ defaultPrevented: false })),
  navigate: jest.fn(),
});

describe('TabBar', () => {
  const mockTabs = [
    { key: 'home', label: 'Home', icon: '🏠' },
    { key: 'profile', label: 'Profile', icon: '👤' },
    { key: 'settings', label: 'Settings', icon: '⚙️' },
  ];

  // Mock insets for safe area
  const mockInsets = { top: 0, right: 0, bottom: 0, left: 0 };

  it('renders all tabs correctly', () => {
    const mockNavigation = createMockNavigation();
    const mockState = {
      routeNames: ['home', 'profile', 'settings'],
      index: 0,
    };

    const { getByText } = render(
      <TabBar 
        state={mockState as any} 
        navigation={mockNavigation as any} 
        tabs={mockTabs}
        descriptors={{} as any}
        insets={mockInsets as any}
      />
    );

    // Check if all tab labels are rendered
    expect(getByText('Home')).toBeTruthy();
    expect(getByText('Profile')).toBeTruthy();
    expect(getByText('Settings')).toBeTruthy();

    // Check if all tab icons are rendered
    expect(getByText('🏠')).toBeTruthy();
    expect(getByText('👤')).toBeTruthy();
    expect(getByText('⚙️')).toBeTruthy();
  });

  it('highlights the active tab', () => {
    const mockNavigation = createMockNavigation();
    const mockState = {
      routeNames: ['home', 'profile', 'settings'],
      index: 1, // Profile tab is active
    };

    const { toJSON } = render(
      <TabBar 
        state={mockState as any} 
        navigation={mockNavigation as any} 
        tabs={mockTabs}
        descriptors={{} as any}
        insets={mockInsets as any}
      />
    );

    // Use snapshot to verify the active tab styling
    expect(toJSON()).toMatchSnapshot();
  });

  it('navigates to the correct tab when pressed', () => {
    const mockNavigation = createMockNavigation();
    const mockState = {
      routeNames: ['home', 'profile', 'settings'],
      index: 0,
    };

    const { getByText } = render(
      <TabBar 
        state={mockState as any} 
        navigation={mockNavigation as any} 
        tabs={mockTabs}
        descriptors={{} as any}
        insets={mockInsets as any}
      />
    );

    // Press the Profile tab
    fireEvent.press(getByText('Profile'));

    // Check if the navigation.emit was called with the correct parameters
    expect(mockNavigation.emit).toHaveBeenCalledWith({
      type: 'tabPress',
      target: 'profile',
      canPreventDefault: true,
    });

    // Check if navigation.navigate was called with the correct route name
    expect(mockNavigation.navigate).toHaveBeenCalledWith('profile');
  });

  it('does not navigate if the event is prevented', () => {
    const mockNavigation = {
      emit: jest.fn(() => ({ defaultPrevented: true })),
      navigate: jest.fn(),
    };
    
    const mockState = {
      routeNames: ['home', 'profile', 'settings'],
      index: 0,
    };

    const { getByText } = render(
      <TabBar 
        state={mockState as any} 
        navigation={mockNavigation as any} 
        tabs={mockTabs}
        descriptors={{} as any}
        insets={mockInsets as any}
      />
    );

    // Press the Settings tab
    fireEvent.press(getByText('Settings'));

    // Check if the navigation.emit was called
    expect(mockNavigation.emit).toHaveBeenCalled();

    // Check that navigation.navigate was NOT called because the event was prevented
    expect(mockNavigation.navigate).not.toHaveBeenCalled();
  });
}); 