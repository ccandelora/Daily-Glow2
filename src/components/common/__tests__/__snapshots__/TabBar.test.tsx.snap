// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabBar highlights the active tab 1`] = `
<View
  style={
    {
      "backgroundColor": "rgba(38, 20, 60, 0.85)",
      "borderTopColor": "rgba(82, 67, 194, 0.3)",
      "borderTopWidth": 1,
      "paddingBottom": 32,
      "paddingTop": 16,
    }
  }
>
  <View
    style={
      {
        "flexDirection": "row",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
          "opacity": 1,
          "paddingHorizontal": 4,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#FFFFFF",
              "fontFamily": "System",
              "fontSize": 16,
            },
            {
              "color": "rgba(255, 255, 255, 0.7)",
              "fontSize": 16,
              "fontWeight": "400",
              "lineHeight": 24,
            },
            {
              "color": "rgba(255, 255, 255, 0.7)",
            },
            null,
            {
              "fontSize": 22,
              "marginBottom": 4,
            },
          ]
        }
      >
        🏠
      </Text>
      <Text
        numberOfLines={1}
        style={
          [
            {
              "color": "#FFFFFF",
              "fontFamily": "System",
              "fontSize": 16,
            },
            {
              "color": "rgba(255, 255, 255, 0.7)",
              "fontSize": 14,
              "fontWeight": "400",
            },
            {
              "color": "rgba(255, 255, 255, 0.7)",
            },
            null,
            undefined,
          ]
        }
      >
        Home
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "flex": 1,
          "justifyContent": "center",
          "opacity": 1,
          "paddingHorizontal": 4,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#FFFFFF",
              "fontFamily": "System",
              "fontSize": 16,
            },
            {
              "color": "rgba(255, 255, 255, 0.7)",
              "fontSize": 16,
              "fontWeight": "400",
              "lineHeight": 24,
            },
            {
              "color": "#00C853",
            },
            null,
            {
              "fontSize": 22,
              "marginBottom": 4,
            },
          ]
        }
      >
        👤
      </Text>
      <Text
        numberOfLines={1}
        style={
          [
            {
              "color": "#FFFFFF",
              "fontFamily": "System",
              "fontSize": 16,
            },
            {
              "color": "rgba(255, 255, 255, 0.7)",
              "fontSize": 14,
              "fontWeight": "400",
            },
            {
              "color": "#00C853",
            },
            null,
            undefined,
          ]
        }
      >
        Profile
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
          "opacity": 1,
          "paddingHorizontal": 4,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#FFFFFF",
              "fontFamily": "System",
              "fontSize": 16,
            },
            {
              "color": "rgba(255, 255, 255, 0.7)",
              "fontSize": 16,
              "fontWeight": "400",
              "lineHeight": 24,
            },
            {
              "color": "rgba(255, 255, 255, 0.7)",
            },
            null,
            {
              "fontSize": 22,
              "marginBottom": 4,
            },
          ]
        }
      >
        ⚙️
      </Text>
      <Text
        numberOfLines={1}
        style={
          [
            {
              "color": "#FFFFFF",
              "fontFamily": "System",
              "fontSize": 16,
            },
            {
              "color": "rgba(255, 255, 255, 0.7)",
              "fontSize": 14,
              "fontWeight": "400",
            },
            {
              "color": "rgba(255, 255, 255, 0.7)",
            },
            null,
            undefined,
          ]
        }
      >
        Settings
      </Text>
    </View>
  </View>
</View>
`;
